presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java

```diff
@@ -34,7 +34,18 @@ default boolean isEnabled(Session session)
         return true;
     }
 
-    Optional<PlanNode> apply(T node, Captures captures, Context context);
+    @Deprecated
+    default Optional<PlanNode> apply(T node, Captures captures, Context context)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    // TODO remove default implementaiton once deprecated apply method got removed
+    default void apply(T node, Captures captures, Context context, Result result)
```

Comment:
I really dislike the pattern of using output arguments to return a result. They make the code harder to reason about, harder to understand concurrency semantics (does something protect the access to those arguments if they need to be mutated?), harder to understand ownership (can I hold on to the object? how about modifying it after return?), etc. Let's not go down that road. I'm fine with apply returning a complex structure if that's necessary.